### DISCUSSION TITLE
Receiving BCC email addresses

### DISCUSSION THREAD

**Discussion by Minisaints** [Jericho/StrongGrid#455] [Q&A]
Hello,

I've noticed that I am receiving BCC emails but they have no recipients. The "WebHook.InboundEmail" class doesn't provide a field for this either? How can I pickup the email from a BCC?

Thanks.

**Comment by Jericho** [comment]
BCC is not a documented parameter (see [here](https://docs.sendgrid.com/for-developers/parsing-email/setting-up-the-inbound-parse-webhook#raw-parameters)), that's why it's not present in StrongGrid's model class for inbound email.

But if you could provide the payload that you receive from SendGrid, I'll be happy to model it in StrongGrid.

**Reply by Minisaints** [reply]
Ah okay that's great. I do believe that it exists in the email headers.

I sent an email from Outlook as "john@external.com" to "peter@internal.com" but only put "peter@internal.com" as a BCC and left CC/To empty. "peter@internal.com" receives the email but StrongGrid returns the "To" field as "john@external.com".

After looking into the email headers through StrongGrid, I can see "From:john@external.com" (shown below) which I'm presuming StrongGrid defaults to if no To/CC exists?

`@From=John <john@external.com>@Subject=BCCTEST-SUBJECT`

 A bit further up in the headers though, SendGrid does send across the actual BCC email which is "peter@internal.com" shown below.

` (Postfix) with ESMTPS id 8E2A45C1810 for <peter@internal.com>`

Hope that makes sense.

**Reply by Jericho** [reply]
What would help would be the actual payload that you receive from SendGrid.

**Reply by Jericho** [reply]
In case it's helpful: this is how I save the payload to a text file when I need to review the data posted by SendGrid

```csharp
private static async Task SaveRequest(HttpRequest request)
{
    const string destinationFolder = @"C:\Temp\";

    try
    {
        // Check if file(s) already exist
        var previousPayloadFiles = Directory.EnumerateFiles(destinationFolder, $"SendGridInboundPayload*.txt").ToArray();

        // Read the payload
        await using var body = new MemoryStream();
        await request.Body.CopyToAsync(body).ConfigureAwait(false);
        body.Position = 0;

        using var reader = new StreamReader(body);
        var postedData = await reader.ReadToEndAsync().ConfigureAwait(false);

        // Save to a file
        var newFileName = Path.Combine(destinationFolder, $"SendGridInboundPayload{(previousPayloadFiles.Length + 1):00}.txt");
        await File.WriteAllTextAsync(newFileName, postedData).ConfigureAwait(false);
    }
    catch (Exception e)
    {
        Debug.WriteLine(e.Message);
    }
}
```

This will allow you to provide the content of the text file

**Reply by Jericho** [reply]
@Minisaints if you still want me to research the possibility of adding BCC, I need you to provide the information I requested.

**Reply by Minisaints** [reply]
Hello, apologises for my absence, had a recent unpleasant experience with COVID. I would still like to research this possibility and I'll try get that information over as soon as possible.

Thanks.

**Comment by Jericho** [comment]
I have done some preliminary testing and, as far as I can tell, the data sent by SendGrid does **not** include a section called "BCC" or something like that. But there's a section called "Envelope" which contains a property called "to" which seems to contain the email address that was BCC'd. This might be the info you are looking for.

I'm still curious to take a look at the data sent to you by SendGrid to verify if they provide you with more info than what my testing has revealed.

**Comment by Jericho** [comment]
@Minisaints are you still interested in researching this situation? Have you looked at the content of the `Envelope.To` property like I previously mentioned? Also, are you able to capture the data posted to you by SendGrid using the helper code I provided?
